HW4_Sapozhnikov (#22)
Python-BI-2023/HW4_Functions2: HW4_Sapozhnikov/README.md

# HW 4. Functions 2
> *This is the repo for the fourth homework of the BI Python 2023 course*

### Prototool
`prototool.py` is a special script for working with polyaminoacid sequences

***

### Overview
`prototool.py` includes 7 methods to treatment of polyaminoacid sequences.

> `prototool.py` can be used for the next goals:

Reviewer: ```suggestion
`prototool.py` can be used for the following purposes:
```

> - recoding 1-letter coded polyaminoacid seqeunces into 3-letter coded and vice versa;

Reviewer: ```suggestion
- recoding 1-letter coded polyaminoacid seqeunces into 3-letter coded and *vice versa*;
```

- polyaminoacid sequences aligment with Smith-Waterman algorithm [^1];
- finding possinle RNA sequences for given polyaminoacid sequences;

> - determining polyaminoacid isoelectric point;
> - calculating polyaminoacid molecular weight;

Reviewer: ```suggestion
- calculating polyaminoacid isoelectric point;
- calculating polyaminoacid molecular weight;
```

- finding possinle DNA sequences for given polyaminoacid sequences; 
- determining GC-content of a corresponding DNA sequence to a given polyaminoacid sequence

***

### Usage
This tool can be used both standalone and as module.
- to use `prototool` standalone you will have to add these lines in the code
  ![image](https://github.com/NSapozhnikov/HW4_Sapozhnikov/assets/81642791/5fa3cf7f-e6f3-4294-9e81-b1ebe17c8514)
  - where *args are sequences you want to process and method is a specified algorithm to use
  - your result will be written in a variable (test on a picture)
- to use `prototool` as module (recomended) you should import it as any other module (check the path: prototools.py should be in the same directory as your script). Then you can freely use any of its functions (see examples).

***

### Options
Arguments:
- `*args[str]` sequences to work with. You can pass several arguments into all functions
- `method` - a method to use

output: All functions return a dict, where keys are original sequenses, values are results after using a corresponding method.

***

### Examples

> def recode allows to translate 1-letter to 3-letters polyaminoacids code

Reviewer: ```suggestion
Function `recode` translates 1-letter to 3-letters polyaminoacids code
```

- `main('AlaValTyr', 'DNT', method = 'recode')`
- `recode('AlaValTyr', 'DNT')`

> - ![image](https://github.com/NSapozhnikov/HW4_Sapozhnikov/assets/81642791/117befa5-feaa-433a-9ac9-23cffe9b024f)

Reviewer: Круто что вы вставили скриншоты! Тем не менее, в реальных тулах и README в таком случае лучше переключаться на светлую тему

> ***

Reviewer: Тут между примерами имхо эти линии не нужны, а то совсем README получается рсчерченым в линеечку

def count_gc_content allows to count the ratio of GC in the entire DNA sequence
- `main('AlaValTyr', 'DNT', method = 'count_gc_content')`
- `count_gc_content('AlaValTyr', 'DNT')`
- ![image](https://github.com/NSapozhnikov/HW4_Sapozhnikov/assets/81642791/d2705714-a3e8-4054-8998-61d922a4feb6)
***

def count_protein_molecular_weight allows to calculate the molecular weight of the polyaminoacid
- `main('AlaValTyr', 'DNT', method = 'count_protein_molecular_weight')`
- `count_protein_molecular_weight('AlaValTyr', 'DNT')`
- ![image](https://github.com/NSapozhnikov/HW4_Sapozhnikov/assets/81642791/cc1eff9a-1b39-4232-98e4-80f622101083)

***

### Troubleshooting
If you have `ValueError("No input defined.")` it means, that you have an empty input. Please, enter the correct input. 
***
If you have `ValueError(method, " is not a valid method.")` it means, that your tool is not correct. Please, enter the right tool.
***
If you have `ValueError('Non-protein aminoacids in sequence')` it means, that your sequences contain non-protein aminoacids. Please, check your sequences and enter the correct input. 

***

### Contributions and contacts

Feel free to report any bugs and problems encountered.

> Email: nikita.sapozhnikov1@gmail.com developed recode(), prettify_alignment(), local_alignmen(), check_input()
> ***
> nekrasovadasha22@mail.ru developed from_proteins_seqs_to_rna(), isoelectric_point_determination()
> *** 
> alina.potyseva@gmail.com developed back_transcribe(), count_gc_content(), count_protein_molecular_weight()

Reviewer: ```suggestion
Nikita Sapozhnikov ([nikita.sapozhnikov1@gmail.com](mailto:nikita.sapozhnikov1@gmail.com)) `developed recode()`, `prettify_alignment()`, `local_alignmen()`, check_input()`
***
Daria Nekrasova ([nekrasovadasha22@mail.ru](mailto:nekrasovadasha22@mail.ru)) developed `from_proteins_seqs_to_rna()`, `isoelectric_point_determination()`
*** 
Potyseva Alina ([alina.potyseva@gmail.com](mailto:alina.potyseva@gmail.com)) developed `back_transcribe()`, `count_gc_content()`, `count_protein_molecular_weight()`
```

***

### References

> [^1]: T.F. Smith, M.S. Waterman, (1981). [Identification of common molecular subsequences](https://doi.org/10.1016/0022-2836(81)90087-5). Journal of Molecular Biology.

Reviewer: 👍
